cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/token/MultiClusterReplicaAwareFailureHandler.java

```diff
@@ -0,0 +1,136 @@
+        {
+            Preconditions.checkState(failureHandlerPerCluster.isEmpty(),
+                                     "Cannot track failures from both instances with and without clusterId");
+            defaultFailureHandler.addFailure(tokenRange, instance, errMessage);
+        }
+    }
+
+    @Override
+    public synchronized Set<I> getFailedInstances()
+    {
+        if (failureHandlerPerCluster.isEmpty())
+        {
+            return defaultFailureHandler.getFailedInstances();
+        }
+
+        // failed instances merged from all clusters
+        HashSet<I> failedInstances = new HashSet<>();
+        failureHandlerPerCluster.values().forEach(handler -> failedInstances.addAll(handler.getFailedInstances()));
+        return failedInstances;
+    }
+
+    @Override
+    public synchronized List<ReplicaAwareFailureHandler<I>.ConsistencyFailurePerRange>
+    getFailedRanges(TokenRangeMapping<I> tokenRangeMapping,
+                    JobInfo job,
+                    ClusterInfo cluster)
+    {
+        if (failureHandlerPerCluster.isEmpty())
+        {
+            return defaultFailureHandler.getFailedRanges(tokenRangeMapping, job, cluster);
+        }
+
+        List<ReplicaAwareFailureHandler<I>.ConsistencyFailurePerRange> failurePerRanges = new ArrayList<>();
+        CoordinatedWriteConf coordinatedWriteConf = job.coordinatedWriteConf();
+        Preconditions.checkState(coordinatedWriteConf != null,
+                                 "CoordinatedWriteConf is absent for multi-cluster write");
+        Preconditions.checkState(cluster instanceof CassandraClusterInfoGroup,
+                                 "Not a CassandraClusterInfoGroup for multi-cluster write");
+        CassandraClusterInfoGroup group = (CassandraClusterInfoGroup) cluster;
+        failureHandlerPerCluster.forEach((clusterId, handler) -> {
+            ClusterConf clusterConf = Objects.requireNonNull(coordinatedWriteConf.cluster(clusterId),
+                                                             () -> "ClusterConf is not found for " + clusterId);
+            ClusterInfo clusterInfo = Objects.requireNonNull(group.cluster(clusterId),
+                                                             () -> "ClusterInfo is not found for " + clusterId);
+            ReplicationFactor rf = clusterInfo.replicationFactor();
+            failurePerRanges.addAll(handler.getFailedRangesInternal(tokenRangeMapping, job.getConsistencyLevel(), clusterConf.localDc(), rf));
+        });
+        return failurePerRanges;
+    }
+
+    @Override
+    protected List<ReplicaAwareFailureHandler<I>.ConsistencyFailurePerRange>
+    getFailedRangesInternal(TokenRangeMapping<I> tokenRangeMapping,
+                            ConsistencyLevel cl,
+                            @Nullable String localDC,
+                            ReplicationFactor replicationFactor)
+    {
+        throw new NotImplementedException("Not implemented for MultiClusterReplicaAwareFailureHandler");
```

Comment:
Can we stick to Java SDK classes?
```suggestion
        throw new UnsupportedOperationException("Not implemented for MultiClusterReplicaAwareFailureHandler");
```